### PR TITLE
add closePeerConnection api to trigger DTLS close.

### PR DESCRIPTION
**Pull request by chehefen**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/source/Dtls/Dtls.c [hunk 1]
```diff
@@ -539,6 +546,7 @@ STATUS dtlsSessionPutApplicationData(PDtlsSession pDtlsSession, PBYTE pData, INT
     SIZE_T pending;
 
     MUTEX_LOCK(pDtlsSession->sslLock);
+    CHK(!ATOMIC_LOAD_BOOL(&pDtlsSession->shutdown), retStatus);
```
Comment (MushMal) [inline src/source/Dtls/Dtls.c:549]: Move this above the lock

File: src/source/Dtls/Dtls.c [hunk 2]
```diff
@@ -561,6 +569,33 @@ STATUS dtlsSessionPutApplicationData(PDtlsSession pDtlsSession, PBYTE pData, INT
...
 
+STATUS dtlsSessionShutdown(PDtlsSession pDtlsSession)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+
+    CHK(pDtlsSession != NULL, STATUS_NULL_ARG);
+
+    MUTEX_LOCK(pDtlsSession->sslLock);
+    CHK(!ATOMIC_LOAD_BOOL(&pDtlsSession->shutdown), retStatus);
```
Comment (MushMal) [inline src/source/Dtls/Dtls.c:580]: Same - move above the lock acquisition

File: src/source/Dtls/Dtls.c [hunk 3]
```diff
@@ -561,6 +569,33 @@ STATUS dtlsSessionPutApplicationData(PDtlsSession pDtlsSession, PBYTE pData, INT
...
+    CHK(ATOMIC_LOAD_BOOL(&pDtlsSession->sslInitFinished), retStatus);
```
Comment (MushMal) [inline src/source/Dtls/Dtls.c:581]: Perhaps this one also needs to be moved above the lock

File: src/source/Dtls/Dtls.c [hunk 4]
```diff
@@ -561,6 +569,33 @@ STATUS dtlsSessionPutApplicationData(PDtlsSession pDtlsSession, PBYTE pData, INT
...
+
+    SSL_shutdown(pDtlsSession->pSsl);
+    CHK_STATUS(dtlsCheckOutgoingDataBuffer(pDtlsSession));
+
+    ATOMIC_STORE_BOOL(&pDtlsSession->shutdown, TRUE);
```
Comment (MushMal) [inline src/source/Dtls/Dtls.c:586]: I think this should go before CHK_STATUS(dtlsCheckOutgoingDataBuffer()) call

File: src/source/Dtls/Dtls.c [hunk 5]
```diff
@@ -561,6 +569,33 @@ STATUS dtlsSessionPutApplicationData(PDtlsSession pDtlsSession, PBYTE pData, INT
...
+
+    CHK_STATUS(dtlsSessionChangeState(pDtlsSession, CLOSED));
```
Comment (MushMal) [inline src/source/Dtls/Dtls.c:588]: Are we calling this API with the lock acquired? This is internal so it should be OK

File: src/source/PeerConnection/PeerConnection.c [hunk 6]
```diff
@@ -1045,6 +1045,25 @@ STATUS restartIce(PRtcPeerConnection pPeerConnection)
...
 
+STATUS closePeerConnection(PRtcPeerConnection pPeerConnection)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PKvsPeerConnection pKvsPeerConnection = (PKvsPeerConnection) pPeerConnection;
+
+    CHK(pKvsPeerConnection != NULL, STATUS_NULL_ARG);
+
+    CHK_STATUS(dtlsSessionShutdown(pKvsPeerConnection->pDtlsSession));
```
Comment (MushMal) [inline src/source/PeerConnection/PeerConnection.c:1056]: Should this be CHK_LOG_ERR?

File: src/source/Dtls/Dtls.c (hunk 1 above)
Comment (chehefen) [reply src/source/Dtls/Dtls.c:549]: I cant, what could happen is the check passes and the thread block at the lock then shutdown happens.

File: src/source/Dtls/Dtls.c (hunk 2 above)
Comment (chehefen) [reply src/source/Dtls/Dtls.c:580]: same reason as before

File: src/source/Dtls/Dtls.c (hunk 3 above)
Comment (chehefen) [reply src/source/Dtls/Dtls.c:581]: same reason.

File: src/source/PeerConnection/PeerConnection.c (hunk 6 above)
Comment (chehefen) [reply src/source/PeerConnection/PeerConnection.c:1056]: sry, iceAgentShutdown should be CHK_STATUS

File: src/source/Dtls/Dtls.c (hunk 1 above)
Comment (MushMal) [reply src/source/Dtls/Dtls.c:549]: I see, no matter how you look at it, it's impossible to properly handle this for all cases. The situation is that if another thread is already shutting down the lock will be released after the shutdown. The main issue is to guard at the higher-layer logic against the object being freed before the check for the atomic.

**Pull request merged**